File: supermarket_system/static/js/inventory.js
const API = '/api/inventory/products';
let productModal;

document.addEventListener('DOMContentLoaded', () => {
    productModal = new bootstrap.Modal(document.getElementById('productModal'));
    loadCategories();
    loadProducts();
});

function showAlert(message, type = 'danger') {
    const alert = document.getElementById('inventoryAlert');
    alert.innerHTML = `<div class="alert alert-${type}" role="alert">${message}</div>`;
    setTimeout(() => { alert.innerHTML = ''; }, 5000);
}

async function loadCategories() {
    try {
        const res = await fetch('/api/inventory/categories');
        const result = await res.json();
        if (!result.success) return;

        const select = document.getElementById('categoryFilter');
        result.data.forEach(cat => {
            const option = document.createElement('option');
            option.value = cat;
            option.textContent = cat;
            select.appendChild(option);
        });
    } catch (err) {
        console.error('Error loading categories:', err);
    }
}

async function loadProducts() {
    try {
        const category = document.getElementById('categoryFilter').value;
        const search = document.getElementById('searchInput').value;
        let url = API + '?page=1&per_page=100';
        if (category) url += `&category=${category}`;
        if (search) url += `&search=${search}`;

        const res = await fetch(url);
        const result = await res.json();
        if (!result.success) return showAlert(result.error || 'Unable to load products');

        const tbody = document.getElementById('productsTableBody');
        tbody.innerHTML = '';
        result.data.forEach(product => {
            const status = product.quantity < product.min_stock ? 
                '<span class="badge bg-warning">Low Stock</span>' : 
                '<span class="badge bg-success">In Stock</span>';
            
            tbody.innerHTML += `
                <tr>
                    <td>${product.sku}</td>
                    <td>${product.name}</td>
                    <td>${product.category}</td>
                    <td>${product.price.toLocaleString('en-US', {minimumFractionDigits: 0})} UGX</td>
                    <td>${product.quantity}</td>
                    <td>${product.min_stock}</td>
                    <td>${status}</td>
                    <td>
                        <button class="btn btn-sm btn-primary me-1" onclick="editProduct(${product.id})">Edit</button>
                        <button class="btn btn-sm btn-danger" onclick="deleteProduct(${product.id})">Delete</button>
                    </td>
                </tr>`;
        });
    } catch (err) {
        showAlert('Error loading products: ' + err.message);
    }
}

function openProductModal() {
    document.getElementById('productId').value = '';
    document.getElementById('productSku').value = '';
    document.getElementById('productName').value = '';
    document.getElementById('productCategory').value = '';
    document.getElementById('productPrice').value = '';
    document.getElementById('productQuantity').value = '0';
    document.getElementById('productMinStock').value = '10';
    document.getElementById('productSupplier').value = '';
    document.getElementById('productDescription').value = '';
    document.getElementById('productActive').checked = true;
    showAlert('');
    productModal.show();
}

async function editProduct(id) {
    try {
        const res = await fetch(`${API}/${id}`);
        const result = await res.json();
        if (!result.success) return showAlert(result.error || 'Unable to load product');

        const product = result.data;
        document.getElementById('productId').value = product.id;
        document.getElementById('productSku').value = product.sku;
        document.getElementById('productName').value = product.name;
        document.getElementById('productCategory').value = product.category;
        document.getElementById('productPrice').value = product.price;
        document.getElementById('productQuantity').value = product.quantity;
        document.getElementById('productMinStock').value = product.min_stock;
        document.getElementById('productSupplier').value = product.supplier || '';
        document.getElementById('productDescription').value = product.description || '';
        document.getElementById('productActive').checked = product.is_active;
        showAlert('');
        productModal.show();
    } catch (err) {
        showAlert('Error fetching product: ' + err.message);
    }
}

async function saveProduct() {
    const id = document.getElementById('productId').value;
    const payload = {
        sku: document.getElementById('productSku').value.trim(),
        name: document.getElementById('productName').value.trim(),
        category: document.getElementById('productCategory').value.trim(),
        price: parseFloat(document.getElementById('productPrice').value),
        quantity: parseInt(document.getElementById('productQuantity').value),
        min_stock: parseInt(document.getElementById('productMinStock').value),
        supplier: document.getElementById('productSupplier').value.trim(),
        description: document.getElementById('productDescription').value.trim(),
        is_active: document.getElementById('productActive').checked
    };

    if (!payload.sku || !payload.name) {
        return showAlert('SKU and name required', 'warning');
    }

    try {
        const method = id ? 'PUT' : 'POST';
        const url = id ? `${API}/${id}` : API;
        const res = await fetch(url, {
            method,
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify(payload)
        });
        const result = await res.json();
        if (!result.success) return showAlert(result.error || 'Error saving product');
        productModal.hide();
        loadProducts();
    } catch (err) {
        showAlert('Error saving product: ' + err.message);
    }
}

async function deleteProduct(id) {
    if (!confirm('Deactivate this product?')) return;
    try {
        const res = await fetch(`${API}/${id}`, { method: 'DELETE' });
        const result = await res.json();
        if (!result.success) return showAlert(result.error || 'Error deleting product');
        loadProducts();
    } catch (err) {
        showAlert('Error deleting product: ' + err.message);
    }
}

File: supermarket_system/modules/inventory.py
from flask import Blueprint, request, jsonify
from models import db, Product
from datetime import datetime
from decimal import Decimal

inventory_bp = Blueprint('inventory', __name__, url_prefix='/api/inventory')

@inventory_bp.route('/products', methods=['GET'])
def list_products():
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 50, type=int)
    category = request.args.get('category', None)
    search = request.args.get('search', None)

    query = Product.query.filter_by(is_active=True)
    
    if category:
        query = query.filter_by(category=category)
    
    if search:
        query = query.filter(Product.name.ilike(f'%{search}%'))
    
    pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    return jsonify({
        'success': True,
        'page': pagination.page,
        'per_page': pagination.per_page,
        'total': pagination.total,
        'data': [p.to_dict() for p in pagination.items]
    }), 200

@inventory_bp.route('/categories', methods=['GET'])
def get_categories():
    categories = db.session.query(Product.category).distinct().all()
    return jsonify({
        'success': True,
        'data': [c[0] for c in categories if c[0]]
    }), 200

@inventory_bp.route('/products', methods=['POST'])
def create_product():
    data = request.get_json() or {}
    name = (data.get('name') or '').strip()
    sku = (data.get('sku') or '').strip()
    
    if not name or not sku:
        return jsonify({'success': False, 'error': 'name and sku required'}), 400

    if Product.query.filter_by(sku=sku).first():
        return jsonify({'success': False, 'error': 'SKU already exists'}), 400

    try:
        product = Product(
            name=name,
            sku=sku,
            category=(data.get('category') or '').strip(),
            price=Decimal(str(data.get('price', 0))),
            quantity=int(data.get('quantity', 0)),
            min_stock=int(data.get('min_stock', 10)),
            supplier=(data.get('supplier') or '').strip(),
            description=(data.get('description') or '').strip(),
            is_active=bool(data.get('is_active', True))
        )
        db.session.add(product)
        db.session.commit()
        return jsonify({'success': True, 'data': product.to_dict()}), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500

@inventory_bp.route('/products/<int:product_id>', methods=['GET'])
def get_product(product_id):
    product = Product.query.get(product_id)
    if not product:
        return jsonify({'success': False, 'error': 'Product not found'}), 404
    return jsonify({'success': True, 'data': product.to_dict()}), 200

@inventory_bp.route('/products/<int:product_id>', methods=['PUT'])
def update_product(product_id):
    product = Product.query.get(product_id)
    if not product:
        return jsonify({'success': False, 'error': 'Product not found'}), 404

    data = request.get_json() or {}
    try:
        if 'name' in data: product.name = (data.get('name') or '').strip()
        if 'sku' in data: 
            sku = (data.get('sku') or '').strip()
            if sku != product.sku and Product.query.filter_by(sku=sku).first():
                return jsonify({'success': False, 'error': 'SKU already exists'}), 400
            product.sku = sku
        if 'category' in data: product.category = (data.get('category') or '').strip()
        if 'price' in data: product.price = Decimal(str(data.get('price')))
        if 'quantity' in data: product.quantity = int(data.get('quantity'))
        if 'min_stock' in data: product.min_stock = int(data.get('min_stock'))
        if 'supplier' in data: product.supplier = (data.get('supplier') or '').strip()
        if 'description' in data: product.description = (data.get('description') or '').strip()
        if 'is_active' in data: product.is_active = bool(data.get('is_active'))

        db.session.commit()
        return jsonify({'success': True, 'data': product.to_dict()}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500

@inventory_bp.route('/products/<int:product_id>', methods=['DELETE'])
def delete_product(product_id):
    product = Product.query.get(product_id)
    if not product:
        return jsonify({'success': False, 'error': 'Product not found'}), 404
    try:
        product.is_active = False
        db.session.commit()
        return jsonify({'success': True, 'message': 'Product deactivated'}), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'error': str(e)}), 500
